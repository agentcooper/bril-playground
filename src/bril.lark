start: (struct | func)*

struct: STRUCT IDENT "=" "{" mbr* "}"
mbr: IDENT ":" type ";"

func: FUNC ["(" arg_list? ")"] [tyann] "{" instr* "}"
arg_list: | arg ("," arg)*
arg: IDENT ":" type
?instr: const | vop | eop | label

const.4: IDENT [tyann] "=" "const" lit ";"
vop.3: IDENT [tyann] "=" op ";"
eop.2: op ";"
label.1: LABEL ":"

op: IDENT (FUNC | LABEL | IDENT)*

?tyann: ":" type

lit: SIGNED_INT  -> int
  | BOOL         -> bool
  | SIGNED_FLOAT -> float
  | "nullptr"    -> nullptr

type: IDENT "<" type ">"  -> paramtype
    | IDENT               -> primtype

BOOL: "true" | "false"
STRUCT: "struct"
IDENT: ("_"|"%"|LETTER) ("_"|"%"|"."|LETTER|DIGIT)*
FUNC: "@" IDENT
LABEL: "." IDENT
COMMENT: /#.*/


%import common.SIGNED_INT
%import common.SIGNED_FLOAT
%import common.WS
%import common.LETTER
%import common.DIGIT
%ignore WS
%ignore COMMENT