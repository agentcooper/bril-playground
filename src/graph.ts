export function postOrder(
  successors: Map<string, Set<string>>,
  entry: string
): string[] {
  const out: string[] = [];
  const explored = new Set<string>();

  function helper(root: string) {
    if (explored.has(root)) {
      return;
    }
    explored.add(root);
    for (const successor of successors.get(root)!) {
      helper(successor);
    }
    out.push(root);
  }

  helper(entry);

  return out;
}
