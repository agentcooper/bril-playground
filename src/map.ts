export function invert<T>(map: Map<T, Set<T>>): Map<T, Set<T>> {
  const result = new Map<T, Set<T>>();
  for (const [key, value] of map.entries()) {
    for (const v of value) {
      if (!result.has(v)) {
        result.set(v, new Set());
      }
      result.get(v)!.add(key);
    }
  }
  return result;
}

export function getWithDefault<K, V>(map: Map<K, V>, key: K, fn: () => V): V {
  if (!map.has(key)) {
    map.set(key, fn());
  }
  return map.get(key)!;
}

export function copy<K, V>(input: Map<K, V[]>): Map<K, V[]> {
  const copy = new Map<K, V[]>();
  for (const [key, value] of input.entries()) {
    copy.set(key, [...value]);
  }
  return copy;
}
