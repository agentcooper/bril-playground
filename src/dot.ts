import { Instruction } from "./bril.ts";
import type { CFG } from "./cfg.ts";

function stringifyOptions(...args: { [key: string]: string }[]) {
  const result = {};
  for (const arg of args) {
    Object.assign(result, arg);
  }
  return (
    "[" +
    Object.entries(result)
      .filter(([_, value]) => Boolean(value))
      .map(([key, value]) => `${key}="${value}"`)
      .join(", ") +
    "]"
  );
}

function getLabel(block: Instruction[]): string {
  return block.map((instruction) => `${stringify(instruction)}\\l`).join("");
}

export function buildDot(cfg: CFG, highlight: Set<string>, current?: string) {
  let nodes = "";
  let edges = "";

  for (const blockName of cfg.blocks.keys()) {
    const block = cfg.getBlockByName(blockName);

    const options = stringifyOptions(
      {
        shape: "box",
        xlabel: blockName,
        label: getLabel(block),
      },
      highlight.has(blockName)
        ? { style: "filled", fillcolor: "aquamarine" }
        : {},
      current === blockName ? { color: "#6c6c6c" } : {}
    );

    nodes += `  "${blockName}" ${options};\n`;

    for (const successorName of cfg.getBlockSuccessors(block)) {
      edges += `  "${blockName}" -> "${successorName}";\n`;
    }
  }

  return `
  digraph main {
    node [style="filled", fillcolor="#cceeff44", fontname = "courier", fontsize="10", color="#dddddd"]
    edge [color="#6c6c6c"]
  ${nodes.trimEnd()}
  
  ${edges.trimEnd()}
  }
      `;
}

export function stringify(instr: any): string {
  if (instr.op === "const") {
    const tyann = instr.type ? `: ${type_to_str(instr.type)}` : "";
    return `${instr.dest ?? ""}${tyann} = const ${
      instr.value?.toString().toLowerCase() ?? ""
    }`;
  } else {
    let rhs = instr.op;
    if (instr.funcs?.length) {
      rhs += ` ${instr.funcs.map((f: any) => `@${f}`).join(" ")}`;
    }
    if (instr.args?.length) {
      rhs += ` ${instr.args.join(" ")}`;
    }
    if (instr.labels?.length) {
      rhs += ` ${instr.labels.map((l: any) => `.${l}`).join(" ")}`;
    }
    if (instr.dest) {
      const tyann = instr.type ? `: ${type_to_str(instr.type)}` : "";
      return `${instr.dest}${tyann} = ${rhs}`;
    } else {
      return rhs;
    }
  }
}

function type_to_str(type: any): string {
  if (typeof type === "object" && type !== null) {
    const [key, value] = Object.entries(type)[0];
    return `${key}<${type_to_str(value)}>`;
  } else {
    return type.toString();
  }
}
