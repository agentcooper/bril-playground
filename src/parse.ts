// @ts-nocheck

// ./generated/bril-parser.js is generated from ./bril.lark using https://github.com/larkjs/lark
import { Program } from "./bril.js";
import { get_parser, Token } from "./generated/bril-parser.js";

Token.prototype.toString = function () {
  return this.value;
};

const _pos = (token) => ({
  line: token.line,
  column: token.column,
});

let transformer = {
  start(items) {
    var structs = items.filter((i) => "mbrs" in i);
    var funcs = items.filter((i) => !("mbrs" in i));
    if (structs.length > 0) {
      return {
        structs: structs,
        functions: funcs,
      };
    } else {
      return {
        functions: funcs,
      };
    }
  },

  func(items) {
    const [name, args, typ] = items;

    const instrs = items.filter(
      (i) => typeof i === "object" && ("op" in i || "label" in i)
    );

    const func = {
      name: name.toString().slice(1), // Strip `@`.
      instrs: instrs,
    };
    if (args.length > 0) {
      func.args = args;
    }
    if (typeof typ === "string") {
      func.type = typ;
    }
    if (this.include_pos) {
      func.pos = _pos(name);
    }
    return func;
  },

  arg(items) {
    const [name, typ] = items;
    return {
      name: name.toString(),
      type: typ,
    };
  },

  struct(items) {
    const [_, name, ...mbrs] = items;
    return {
      name: name,
      mbrs: mbrs,
    };
  },

  mbr(items) {
    const [name, typ] = items;
    return {
      name: name,
      type: typ,
    };
  },

  arg_list(items) {
    return items;
  },

  const(items) {
    const [dest, type, val] = items;
    const out = {
      op: "const",
      dest: dest.toString(),
      value: val,
    };
    if (type) {
      out.type = type;
    }
    if (this.include_pos) {
      out.pos = _pos(dest);
    }
    return out;
  },

  vop(items) {
    const [dest, type, op] = items;
    const out = { dest: dest.toString() };
    if (type) {
      out.type = type;
    }
    Object.assign(out, op);
    if (this.include_pos) {
      out.pos = _pos(dest);
    }
    return out;
  },

  op(items) {
    const op_token = items.shift();
    const opcode = op_token.toString();

    const funcs = [];
    const labels = [];
    const args = [];
    for (const item of items) {
      if (item.type === "FUNC") {
        funcs.push(item.toString().slice(1));
      } else if (item.type === "LABEL") {
        labels.push(item.toString().slice(1));
      } else {
        args.push(item.toString());
      }
    }

    const out = { op: opcode };
    if (args.length) {
      out.args = args;
    }
    if (funcs.length) {
      out.funcs = funcs;
    }
    if (labels.length) {
      out.labels = labels;
    }
    if (this.include_pos) {
      out.pos = _pos(op_token);
    }
    return out;
  },

  eop(items) {
    const [op] = items;
    return op;
  },

  label(items) {
    const name = items[0];
    const out = { label: String(name).slice(1) }; // Strip `.`.
    if (this.include_pos) {
      out.pos = _pos(name);
    }
    return out;
  },

  int(items) {
    return parseInt(String(items[0]));
  },

  bool(items) {
    if (String(items[0]) === "true") {
      return true;
    } else {
      return false;
    }
  },

  paramtype(items) {
    return { [items[0]]: items[1] };
  },

  primtype(items) {
    return String(items[0]);
  },

  float(items) {
    return parseFloat(items[0]);
  },

  nullptr(items) {
    return 0;
  },
};

const parser = get_parser({ transformer });

export function parse(input: string): Program {
  return parser.parse(input);
}
