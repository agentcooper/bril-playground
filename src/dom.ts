import { invert } from "./map.ts";

export function dominatorTree(
  dom: Map<string, Set<string>>
): Map<string, Set<string>> {
  const dom_inv = invert(dom);
  const dom_inv_strict = new Map<string, Set<string>>();
  for (const [a, bs] of dom_inv.entries()) {
    const strictBs = new Set(bs);
    strictBs.delete(a);
    dom_inv_strict.set(a, strictBs);
  }

  const dom_inv_strict_2x = new Map<string, Set<string>>();
  for (const [a, bs] of dom_inv_strict.entries()) {
    const resultBs = new Set<string>();
    for (const b of bs) {
      for (const bb of dom_inv_strict.get(b)!) {
        resultBs.add(bb);
      }
    }
    dom_inv_strict_2x.set(a, resultBs);
  }

  const result = new Map<string, Set<string>>();
  for (const [a, bs] of dom_inv_strict.entries()) {
    const resultBs = new Set<string>();
    for (const b of bs) {
      if (!dom_inv_strict_2x.get(a)!.has(b)) {
        resultBs.add(b);
      }
    }
    result.set(a, resultBs);
  }
  return result;
}
