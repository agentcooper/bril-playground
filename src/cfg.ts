import type * as bril from "./bril.ts";

import { dominatorTree } from "./dom.ts";
import { postOrder } from "./graph.ts";
import { getWithDefault, invert } from "./map.ts";
import { equal, filter, intersect, map } from "./set.ts";
import { getTypes, insert_phis, ssa_rename } from "./ssa.ts";

type Block = bril.Instruction[];

const terminators = new Set(["br", "jmp", "ret"]);

function isTerminator(instr: bril.Instruction) {
  return terminators.has(instr.op);
}

function isLabel(instr: bril.Instruction | bril.Label): instr is bril.Label {
  return "label" in instr;
}

function lastInstruction(block: Block) {
  return block[block.length - 1];
}

export class CFG {
  fn: bril.Function;
  blocks: Map<string, Block>;

  constructor(fn: bril.Function) {
    const blocks = this.getBasicBlocks(fn);
    const map = new Map<string, Block>();
    for (let block of blocks) {
      let name: string;
      const [firstInstruction, ...rest] = block;
      if (isLabel(firstInstruction)) {
        name = firstInstruction.label;
        block = rest;
      } else {
        name = this.generateName(`bb`, new Set(map.keys()));
      }
      map.set(name, block as Block);
    }
    this.blocks = map;
    this.addEntry();
    this.addTerminators();
    this.fn = fn;
  }

  private getBasicBlocks(fn: bril.Function) {
    let currentBlock: Array<bril.Instruction | bril.Label> = [];
    const blocks: Array<typeof currentBlock> = [];
    for (const instruction of fn.instrs) {
      if (isLabel(instruction)) {
        if (currentBlock.length > 0) {
          blocks.push(currentBlock);
        }
        currentBlock = [instruction];
      } else {
        currentBlock.push(instruction);
        if (isTerminator(instruction)) {
          blocks.push(currentBlock);
          currentBlock = [];
        }
      }
    }
    if (currentBlock.length > 0) {
      blocks.push(currentBlock);
    }
    return blocks;
  }

  getBlockSuccessors(block: bril.Instruction[]): Set<string> {
    return this.getSuccessors(lastInstruction(block));
  }

  getSuccessors(instruction: bril.Instruction): Set<string> {
    if (instruction.op === "br" || instruction.op === "jmp") {
      return new Set(instruction.labels);
    }
    return new Set();
  }

  getBlockByName(name: string): Block {
    return this.blocks.get(name)!;
  }

  getEdges(): [Map<string, Set<string>>, Map<string, Set<string>>] {
    const successors = new Map<string, Set<string>>();
    const predecessors = new Map<string, Set<string>>();
    for (const blockName of this.blocks.keys()) {
      successors.set(blockName, new Set());
      predecessors.set(blockName, new Set());
    }
    for (const [blockName, block] of this.blocks.entries()) {
      for (const successorName of this.getBlockSuccessors(block)) {
        successors.get(blockName)!.add(successorName);
        predecessors.get(successorName)!.add(blockName);
      }
    }
    return [successors, predecessors];
  }

  reassemble(): (bril.Instruction | bril.Label)[] {
    const instructions: (bril.Instruction | bril.Label)[] = [];
    for (const [blockName, block] of this.blocks.entries()) {
      instructions.push({ label: blockName });
      for (const instruction of block) {
        instructions.push(instruction);
      }
    }
    return instructions;
  }

  getPhis(): Map<string, Set<string>> {
    const df = this.computeDominatorFrontiers();
    const defs = this.computeDefinitions();
    const phis = new Map<string, Set<string>>();
    for (const blockName of this.blocks.keys()) {
      phis.set(blockName, new Set());
    }
    for (const [v, definitionBlocks] of defs) {
      const defBloks = [...new Set(definitionBlocks)];
      for (const def of defBloks) {
        for (const b of df.get(def)!) {
          if (!phis.get(b)!.has(v)) {
            phis.get(b)!.add(v);
            if (!defBloks.includes(b)) {
              defBloks.push(b);
            }
          }
        }
      }
    }
    return phis;
  }

  computeDefinitions(): Map<string, Set<string>> {
    const result = new Map<string, Set<string>>();
    for (const [blockName, block] of this.blocks.entries()) {
      for (const instruction of block) {
        if ("dest" in instruction) {
          getWithDefault(result, instruction.dest, () => new Set()).add(
            blockName
          );
        }
      }
    }
    return result;
  }

  /**
   * Block A dominates block B if every path from entry block to block B includes A.
   *
   * Means that A must always execute before B.
   *
   * @returns Map of dominators for each block.
   */
  computeDominators(): Map<string, Set<string>> {
    const entry = [...this.blocks.keys()][0];
    const [successors, predecessors] = this.getEdges();
    const dom = new Map<string, Set<string>>();
    for (const blockName of this.blocks.keys()) {
      dom.set(blockName, new Set(this.blocks.keys()));
    }
    const reversePostOrder = [...postOrder(successors, entry)].reverse();
    while (true) {
      let changed = false;
      for (const blockName of reversePostOrder) {
        const newDom = intersect(
          Array.from(map(predecessors.get(blockName)!, (p) => dom.get(p)!))
        );
        newDom.add(blockName);
        if (!equal(dom.get(blockName)!, newDom)) {
          dom.set(blockName, newDom);
          changed = true;
        }
      }
      if (!changed) {
        break;
      }
    }
    return dom;
  }

  computeDominatorFrontiers(): Map<string, Set<string>> {
    const [successors] = this.getEdges();
    const dominatedBy = this.computeDominators();
    const dominates = invert(dominatedBy);
    const frontiers = new Map<string, Set<string>>();
    for (const block of this.blocks.keys()) {
      frontiers.set(block, new Set());
    }
    for (const block of dominatedBy.keys()) {
      const dominated_succs = new Set<string>();
      for (const domi of dominates.get(block)!) {
        for (const s of successors.get(domi)!) {
          dominated_succs.add(s);
        }
      }
      frontiers.set(
        block,
        filter(
          dominated_succs,
          (k) => !dominates.get(block)!.has(k) || k === block
        )
      );
    }
    return frontiers;
  }

  addEntry(): void {
    const firstLabel = [...this.blocks.keys()][0];
    let found = false;
    for (const block of this.blocks.values()) {
      for (const instruction of block) {
        if (
          "labels" in instruction &&
          instruction.labels?.includes(firstLabel)
        ) {
          found = true;
          break;
        }
      }
    }
    if (!found) {
      return;
    }
    const entry = this.generateName(`entry`, new Set(this.blocks.keys()));
    // reinsert keys so that entry comes first
    const copy = new Map(this.blocks);
    this.blocks.clear();
    this.blocks.set(entry, []);
    for (const [key, value] of copy) {
      this.blocks.set(key, value);
    }
  }

  addTerminators(): void {
    const blockNames = [...this.blocks.keys()];
    for (let i = 0; i < blockNames.length; i++) {
      const isLast = i == blockNames.length - 1;
      const blockName = blockNames[i];
      const block = this.blocks.get(blockName)!;
      if (block.length === 0 || !isTerminator(lastInstruction(block))) {
        if (isLast) {
          block.push({ op: "ret", args: [] });
        } else {
          block.push({ op: "jmp", labels: [blockNames[i + 1]] });
        }
      }
    }
  }

  addSSA(): void {
    const domTree = dominatorTree(this.computeDominators());
    const phis = this.getPhis();
    const [successors] = this.getEdges();
    const arg_names = (this.fn.args ?? []).map((arg) => arg.name);
    const ssa_result = ssa_rename(
      this.blocks,
      phis,
      successors,
      domTree,
      arg_names
    );
    insert_phis(this.blocks, ssa_result, getTypes(this.fn));
  }

  private generateName(prefix: string, usedNames: Set<string>): string {
    let counter = 1;
    while (true) {
      const candidate = `${prefix}${counter}`;
      if (!usedNames.has(candidate)) {
        return candidate;
      }
      counter += 1;
    }
  }
}
