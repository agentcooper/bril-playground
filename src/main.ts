import * as d3 from "d3";
import { graphviz } from "d3-graphviz";
import { EditorView, basicSetup } from "codemirror";

import { CFG } from "./cfg.ts";
import { buildDot } from "./dot.ts";
import { parse } from "./parse";
import { Program } from "./bril";
import { evalProg } from "./interpreter.ts";
import { invert } from "./map.ts";

enum Hover {
  dominators,
  dominatedBy,
  frontiers,
}

interface State {
  convertToSSA: boolean;
  current?: string;
  hover: Set<string>;
  cfgName: string;
  cfg?: CFG;
  dominators: Map<string, Set<string>>;
  dominatedBy: Map<string, Set<string>>;
  frontiers: Map<string, Set<string>>;
  onHover: Hover;
}

const state: State = {
  convertToSSA: false,
  current: undefined,
  hover: new Set(),
  cfgName: "main",
  cfg: undefined,
  dominators: new Map(),
  dominatedBy: new Map(),
  frontiers: new Map(),
  onHover: Hover.dominators,
};

function compute() {
  const source = editor.state.doc.toString();
  const program: Program = parse(source);

  const fn =
    program.functions.find((f) => f.name === state.cfgName) ??
    program.functions[0];
  const cfg = new CFG(fn);

  if (state.convertToSSA) {
    cfg.addSSA();
  }

  state.cfg = cfg;
  state.dominators = cfg.computeDominators();
  state.dominatedBy = invert(state.dominators);
  state.frontiers = cfg.computeDominatorFrontiers();
}

function render() {
  compute();

  if (!state.cfg) {
    return;
  }

  const dot = buildDot(state.cfg, state.hover, state.current);

  graphviz("#graph", {
    useWorker: false, // could not figure out the proper way to use worker with Vite
  })
    .zoom(false)
    .renderDot(dot)
    .on("end", interactive);
}

function onHover(blockName: string): Set<string> {
  switch (state.onHover) {
    case Hover.dominators:
      return state.dominators.get(blockName)!;
    case Hover.dominatedBy:
      return state.dominatedBy.get(blockName)!;
    case Hover.frontiers:
      return state.frontiers.get(blockName)!;
  }
}

function interactive() {
  d3.selectAll(".node")
    .on("mouseover", (_event, data) => {
      const blockName: string = (data as unknown as { key: string }).key;
      state.current = blockName;
      state.hover = onHover(blockName);
      render();
    })
    .on("mouseout", () => {
      state.current = undefined;
      state.hover = new Set();
      render();
    });
}

const sourceSelect = document.getElementById("source-select")!;

let updateListenerExtension = EditorView.updateListener.of((update) => {
  if (update.docChanged) {
    render();
  }
});

let editor = new EditorView({
  extensions: [basicSetup, updateListenerExtension],
  parent: document.getElementById("editor")!,
});

sourceSelect.addEventListener(
  "change",
  async (event) => {
    state.cfgName = "main";
    const fileName = (event.target as HTMLSelectElement).value;
    const url = `${fileName}`;
    const res = await fetch(url);
    const text = await res.text();
    editor.dispatch({
      changes: {
        from: 0,
        to: editor.state.doc.length,
        insert: text,
      },
    });
  },
  false
);

document
  .getElementById("ssa-checkbox")!
  .addEventListener("change", async (event) => {
    state.convertToSSA = (event.target as HTMLInputElement).checked;
    render();
  });

document
  .getElementById("cfg-function-name")!
  .addEventListener("input", async (event) => {
    state.cfgName = (event.target as HTMLInputElement).value;
    render();
  });

function toHover(input: string): Hover {
  switch (input) {
    case "dominators":
      return Hover.dominators;
    case "dominatedBy":
      return Hover.dominatedBy;
    case "frontiers":
      return Hover.frontiers;
    default:
      throw new Error(`Unknown value: ${input}`);
  }
}

document.getElementById("on-hover-select")!.addEventListener(
  "change",
  async (event) => {
    state.onHover = toHover((event.target as HTMLSelectElement).value);
    render();
  },
  false
);

function getRunArguments() {
  const node = document.getElementById(
    "run-arguments-input"
  ) as HTMLInputElement;

  if (!node.value) {
    return [];
  }

  return node.value.split(",");
}

document.getElementById("run-button")!.addEventListener("click", async () => {
  const source = editor.state.doc.toString();
  const program: Program = parse(source);
  evalProg(program, getRunArguments(), false);
});

setTimeout(() => {
  sourceSelect.dispatchEvent(new Event("change"));
}, 0);
