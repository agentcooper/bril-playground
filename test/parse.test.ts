import { expect, test } from "vitest";
import { readFileSync } from "fs";
import * as path from "path";
import { parse } from "../src/parse";

test("parse", () => {
  const a = readFileSync(
    path.join(__dirname, "../public/condition.bril")
  ).toString();

  expect(parse(a)).toMatchSnapshot();
});
