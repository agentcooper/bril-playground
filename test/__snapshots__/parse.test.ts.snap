// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse 1`] = `
{
  "functions": [
    {
      "instrs": [
        {
          "dest": "a",
          "op": "const",
          "type": "int",
          "value": 47,
        },
        {
          "dest": "b",
          "op": "const",
          "type": "int",
          "value": 42,
        },
        {
          "dest": "cond",
          "op": "const",
          "type": "bool",
          "value": true,
        },
        {
          "args": [
            "cond",
          ],
          "labels": [
            "left",
            "right",
          ],
          "op": "br",
        },
        {
          "label": "left",
        },
        {
          "dest": "b",
          "op": "const",
          "type": "int",
          "value": 1,
        },
        {
          "dest": "c",
          "op": "const",
          "type": "int",
          "value": 5,
        },
        {
          "labels": [
            "end",
          ],
          "op": "jmp",
        },
        {
          "label": "right",
        },
        {
          "dest": "a",
          "op": "const",
          "type": "int",
          "value": 2,
        },
        {
          "dest": "c",
          "op": "const",
          "type": "int",
          "value": 10,
        },
        {
          "labels": [
            "end",
          ],
          "op": "jmp",
        },
        {
          "label": "end",
        },
        {
          "args": [
            "a",
            "c",
          ],
          "dest": "d",
          "op": "sub",
          "type": "int",
        },
        {
          "args": [
            "d",
          ],
          "op": "print",
        },
      ],
      "name": "main",
    },
  ],
}
`;
